library/ssl_tls.c

```diff
@@ -4981,25 +4981,28 @@ static const int ssl_preset_suiteb_ciphersuites[] = {
  */
 static uint16_t ssl_preset_default_sig_algs[] = {
 
-#if defined(MBEDTLS_PK_CAN_ECDSA_SOME) &&  \
+#if defined(MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED) &&  \
```

Comment:
I think we also want MBEDTLS_KEY_EXCHANGE_ECDSA_CERT_REQ_ALLOWED_ENABLED here as this can be used for client auth. (Please correct me if I'm wrong!)

Edit: same for all `preset` variables below.

Reply:
@valeriosetti I am a bit confused here over what has been implemented here. I was expecting to see `MBEDTLS_KEY_EXCHANGE_ECDSA_CERT_REQ_ALLOWED_ENABLED` here. Did you come to the conclusion that is not warranted or have I missed something?

Reply:
TBH the goal of this PR is not easy also for me, so I might be wrong, but here's my idea behind that.

`MBEDTLS_KEY_EXCHANGE_ECDSA_CERT_REQ_ALLOWED_ENABLED` refers to TLS 1.2 key exchanges, while `MBEDTLS_KEY_EXCHANGE_ECDSA_CERT_REQ_ANY_ALLOWED_ENABLED` would be more appropriate for referring to both TLS 1.2 and 1.3. However, since the symbols defined in this array concern TLS 1.3, I thought that just keeping the TLS 1.3 part of `MBEDTLS_KEY_EXCHANGE_ECDSA_CERT_REQ_ANY_ALLOWED_ENABLED` (i.e. `MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED`) could be enough.

The CI seems to agree with that, but as I'm not an expert at all of this, so my conclusion might be wrong. Any suggestion is accepted on this :)

Reply:
Yeah, what your saying does make sense to me. I am also not an expert and this stuff is confusing. I will have another think about it and re-review.

Reply:
> However, since the symbols defined in this array concern TLS 1.3

Ah, but that's the tricky part: even though those symbols have TLS 1.3 in their name, they're not just for 1.3: actually `MBEDTLS_TLS1_3_SIG_ECDSA_SECP256R1_SHA256` can also be spelt `MBEDTLS_SSL_TLS12_SIG_AND_HASH_ALG(MBEDTLS_SSL_SIG_ECDSA, MBEDTLS_SSL_HASH_SHA512)` and then it suddently doesn't looks that specific to TLS 1.3 any more :) (It's not just a numerical coincidence, it was done on purpose, see [RFC 8446 section 4.2.3](https://datatracker.ietf.org/doc/html/rfc8446#section-4.2.3), last-but-one paragraph: "ECDSA signature schemes align with TLS 1.2's ECDSA hash/signature pairs."

Coming back to our code, the comment just above this says that `ssl_preset` "is for TLS 1.3 only or hybrid TLS 1.3/1.2 handshakes", so for algorithms that are common with TLS 1.2, we need to use a `ANY` macro, not a `TLS1_3` macro.

Also, for TLS 1.3, we shouldn't remove the `PK_CAN_ECDSA` condition: we we can't do ECDSA, we shouldn't advertise it.